File: app/src/main/cpp/wn-steam-client/src/pb/cmsg_client_request_encrypted_app_ticket.cpp
#include "wn_steam/pb/cmsg_client_request_encrypted_app_ticket.h"

#include "wn_steam/proto_wire.h"

namespace wn_steam::pb {

std::vector<uint8_t> CMsgClientRequestEncryptedAppTicket::serialize() const {
    std::vector<uint8_t> out;
    proto::Writer w(out);
    w.uint32_field(1, app_id);
    return out;
}

std::optional<CMsgClientRequestEncryptedAppTicketResponse>
CMsgClientRequestEncryptedAppTicketResponse::deserialize(
        std::span<const uint8_t> body) noexcept {
    proto::Reader r(body);
    CMsgClientRequestEncryptedAppTicketResponse m;
    while (!r.eof()) {
        auto t = r.next_tag();
        if (!t) {
            if (!r.ok()) return std::nullopt;
            break;
        }
        switch (t->field_number) {
            case 1:
                if (auto v = r.u32(); v) m.app_id = *v;
                else return std::nullopt;
                break;
            case 2:
                if (auto v = r.i32(); v) m.eresult = *v;
                else return std::nullopt;
                break;
            case 3: {
                // Raw serialized EncryptedAppTicket sub-message.
                auto b = r.bytes();
                if (!b) return std::nullopt;
                m.encrypted_app_ticket.assign(b->begin(), b->end());
                break;
            }
            default:
                if (!r.skip(t->wire_type)) return std::nullopt;
                break;
        }
    }
    return m;
}

}  // namespace wn_steam::pb

File: app/src/main/cpp/wn-steam-client/rust/src/pb/cmsg_client_request_encrypted_app_ticket.rs
use crate::proto_wire::{Reader, Writer};

#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
pub struct CMsgClientRequestEncryptedAppTicket {
    pub app_id: u32,
}

impl CMsgClientRequestEncryptedAppTicket {
    pub fn serialize(&self) -> Vec<u8> {
        let mut out = Vec::new();
        Writer::new(&mut out).uint32_field(1, self.app_id);
        out
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct CMsgClientRequestEncryptedAppTicketResponse {
    pub app_id: u32,
    pub eresult: i32,
    pub encrypted_app_ticket: Vec<u8>,
}

impl Default for CMsgClientRequestEncryptedAppTicketResponse {
    fn default() -> Self {
        Self {
            app_id: 0,
            eresult: 2,
            encrypted_app_ticket: Vec::new(),
        }
    }
}

impl CMsgClientRequestEncryptedAppTicketResponse {
    pub fn deserialize(body: &[u8]) -> Option<Self> {
        let mut reader = Reader::new(body);
        let mut msg = Self::default();
        while !reader.eof() {
            let Some(tag) = reader.next_tag() else {
                return reader.ok().then_some(msg);
            };
            match tag.field_number {
                1 => msg.app_id = reader.u32()?,
                2 => msg.eresult = reader.i32()?,
                3 => msg.encrypted_app_ticket = reader.bytes()?.to_vec(),
                _ => {
                    if !reader.skip(tag.wire_type) {
                        return None;
                    }
                }
            }
        }
        Some(msg)
    }
}
